internal/libs/protoio/writer.go

```diff
@@ -90,11 +91,44 @@ func (w *varintWriter) Close() error {
 	return nil
 }
 
+func varintWrittenBytes(m marshaler, size int) ([]byte, error) {
+	buf := make([]byte, size+binary.MaxVarintLen64)
+	n := binary.PutUvarint(buf, uint64(size))
+	nw, err := m.MarshalTo(buf[n:])
+	if err != nil {
+		return nil, err
+	}
+	return buf[:n+nw], nil
+}
+
+var bufPool = &sync.Pool{
+	New: func() interface{} {
+		return new(bytes.Buffer)
+	},
+}
```

Comment:
I was skeptical that the pool was buying us much here, so I benchmarked this change with ("old") and without ("new") the pool, i.e., just allocating a fresh buffer in the fallback case. It makes no difference in performance, and the code is simpler:

```
name                                                                old time/op    new time/op    delta
VoteSignBytes-12                                                       558ns ± 0%     556ns ± 0%  -0.54%
CommitVoteSignBytes-12                                                6.81µs ± 0%    6.65µs ± 0%  -2.39%

name                                                                old alloc/op   new alloc/op   delta
VoteSignBytes-12                                                        600B ± 0%      600B ± 0%   0.00%
CommitVoteSignBytes-12                                                7.60kB ± 0%    7.60kB ± 0%   0.00%

name                                                                old allocs/op  new allocs/op  delta
VoteSignBytes-12                                                        10.0 ± 0%      10.0 ± 0%   0.00%
CommitVoteSignBytes-12                                                   110 ± 0%       110 ± 0%   0.00%
```

The "new" branch is with the suggestions I made below.

```suggestion
```

Reply:
It absolutely does make a difference which is the point of this PR when you have code that takes the next path where it doesn't implement MarshalTo. You need to benchmark with protobuf structs that doesn't have a MarshalTo method.

Reply:
I just ran the benchmarks you checked in. Which benchmarks are you suggesting I should compare to? I'm not suggesting you remove the special case—only how the buffer is allocated.

Reply:
Note also that allocating a default-size byte buffer (which is what the zero Buffer does) is many times faster than fetching from a sync.Pool, even if you always have a value ready (<1ns vs. ~10ns). In this case it doesn't actually make a meaningful difference, but if you really want to shave cycles, the pool isn't helping at all.

Reply:
Perhaps I have not been clear in my comments, please help me clarify if I'm missing something.

The PR description gives benchmark results showing performance improvements due to the changes in this PR. Those same benchmarks _without_ part (1) of what you're describing—but preserving (2) and (3)—do not show any improvement in performance.

If those benchmarks do not exercise the change sufficiently to distinguish the case where a value does not have a MarshalTo method, which benchmark do you suggest I run to verify that the pool actually helps?

Reply:
In your tests, comment out the path in which I detect if the object implements Marshaler then make it take the path of using the buffer pool always then run the benchmarks before and after and that'll give you the answer for the clear performance improvements.

Reply:
I added that change because there will be structs say from other protobuf libraries that we don't control that'll go through here. Committing such code and generating those structs is overkill and unnecessary for this change, but I've detailed how we can see those improvements.

Reply:
> In your tests, comment out the path in which I detect if the object implements Marshaler then make it take the path of using the buffer pool always then run the benchmarks before and after and that'll give you the answer for the clear performance improvements.

If that path is meaningfully different, the benchmarks should exercise it without the need to manually comment out sections of the implementation. It's possible, but would be very surprising, if a sync.Pool outperforms an explicit allocation of a bytes.Buffer on the cold path. 

I'd expect the "hot" path (the MarshalTo) case to provide most or all of the performance benefit here—and that seems fine. If there's an additional increment from pooling buffers, the tests need to show it.

Reply:
The reasoning I mentioned before is that we would have to generate code using another protobuf generator then commit it in here, we'd have to write a .proto file, make a Makefile entry, have generated code checked in: this is overkill. Moreover the library is meant to handle all paths, yet sadly never benchmarked for many years until I noticed allocations during continuous profiling, and then this path by auditing it.

I implore you to try out and see the actual results with the simple experiment above simply or write a small experiment isolated using a sync.Pool and buffer.
If you aren't able to run that experiment I showed above (but should be a simple one), I'd say trust me, I've done the sync.Pool and bytes.Buffer optimization tones of times and it is a well used allocation reducer even in the Go standard library.

Reply:
> The reasoning I mentioned before is that we would have to generate code using another protobuf generator then commit it in here,

That seems fine. We don't need to check in the schema, any arbitrary type will do. The point is to make the results reproducible in the future, when the current context has been forgotten. It should be fine to define a single message type, generate its code, and copy that code into the benchmark. After all, the only properties that matter are that it should be reachable by `proto.Marshal` and yet not have a `MarshalTo` method.

I do not ask for this from lack of faith in your abilities, but to ensure we have a clear record of why we're doing them. Otherwise, a little bit of extra work now is avoided at the cost of somebody else's confusion later.

Reply:
What I was saying is that it's about us not being pedantic and also being careful in examining the code, you were suggesting a revert with conviction but didn't examine the explanation I provided a few times; I showed how you could trivially examine the performance benefits without bloating the code. Anyways I've added benchmarks and the whole nine yards but that was honestly unnecessary. Thanks for taking a look and for the discourse.

Reply:
Thank you for adding the benchmarks, I appreciate the effort.  These verify what I expected, namely, that the sync.Pool adds no significant improvement compared to allocating the buffer in the fallback case.

```
name                              old time/op    new time/op    delta
MarshalDelimitedWithMarshalTo-12     716ns ± 0%     716ns ± 0%   ~     (p=1.000 n=1+1)
MarshalDelimitedNoMarshalTo-12       783ns ± 0%     868ns ± 0%   ~     (p=1.000 n=1+1)

name                              old alloc/op   new alloc/op   delta
MarshalDelimitedWithMarshalTo-12      360B ± 0%      360B ± 0%   ~     (all equal)
MarshalDelimitedNoMarshalTo-12        440B ± 0%      680B ± 0%   ~     (p=1.000 n=1+1)

name                              old allocs/op  new allocs/op  delta
MarshalDelimitedWithMarshalTo-12      6.00 ± 0%      6.00 ± 0%   ~     (all equal)
MarshalDelimitedNoMarshalTo-12        7.00 ± 0%      9.00 ± 0%   ~     (p=1.000 n=1+1)
```

That said, I'm fine with the implementation you have here, it's not making things any worse than the simpler case.

Reply:
I don't follow what benchmarks you are using but in my results, there is a clear distinction from before and after for NoMarshalTo with the benchmarks I showed and included in my results, otherwise I'd not get any deltas for that case. Also using n=1 for benchmarks is guaranteed to give you bad results (that's using a sample size of 1), you always need like n>10 to get proper results given that benchstat performs statistical sampling

```shell
$ benchstat before.txt after.txt
name                                        old time/op    new time/op    delta
types.VoteSignBytes-8                       705ns ± 3%     573ns ± 6%       -18.74% (p=0.000 n=18+20)
types.CommitVoteSignBytes-8                 8.15µs ± 9%    6.81µs ± 4%      -16.51% (p=0.000 n=20+19)
protoio.MarshalDelimitedWithMarshalTo-8     788ns ± 8%     772ns ± 3%       -2.01%  (p=0.050 n=20+20)
protoio.MarshalDelimitedNoMarshalTo-8       989ns ± 4%     845ns ± 2%       -14.51% (p=0.000 n=20+18)

name                                        old alloc/op   new alloc/op    delta
types.VoteSignBytes-8                       792B ± 0%      600B ± 0%       -24.24%  (p=0.000 n=20+20)
types.CommitVoteSignBytes-8                 9.52kB ± 0%    7.60kB ± 0%     -20.17%  (p=0.000 n=20+20)
protoio.MarshalDelimitedNoMarshalTo-8       808B ± 0%      440B ± 0%       -45.54%  (p=0.000 n=20+20)

name                                        old allocs/op  new allocs/op   delta
types.VoteSignBytes-8                       13.0 ± 0%      10.0 ± 0%       -23.08%  (p=0.000 n=20+20)
types.CommitVoteSignBytes-8                 140 ± 0%       110 ± 0%        -21.43%  (p=0.000 n=20+20)
protoio.MarshalDelimitedNoMarshalTo-8       10.0 ± 0%      7.0 ± 0%        -30.00%  (p=0.000 n=20+20)
```